connector_openproject/components/importer.py

```
@@ -0,0 +1,408 @@
+    _name = "openproject.account.analytic.line.importer"
+    _inherit = "base.openproject.importer"
+    _apply_on = "openproject.account.analytic.line"
+
+    def import_dependencies(self, record):
+        self._import_link_dependency(record, OP_WORK_PACKAGE_LINK, raise_null=False)
+        self._import_link_dependency(record, OP_USER_LINK)
+
+
+class OpenProjectMailMessageImporter(Component):
+    _name = "openproject.mail.message.importer"
+    _inherit = "base.openproject.importer"
+    _apply_on = "openproject.mail.message"
+
+    def _should_skip(self, record):
+        wp_id_, wp = self._link_to_internal(record, OP_WORK_PACKAGE_LINK)
+        project = self.env["openproject.project.project"].search(
+            [
+                ("odoo_id", "=", wp.project_id.id),
+            ]
+        )
+        enabled_for = project.sync_activities
+        activity_type = record["_type"]
+        if should_skip_activity(activity_type, enabled_for):
+            return f"Skipping sync for activity type: {activity_type}"
+
+    def import_dependencies(self, record):
+        self._import_link_dependency(record, OP_WORK_PACKAGE_LINK)
+        self._import_link_dependency(record, OP_USER_LINK)
+
+
+class ImageImporter(AbstractComponent):
+    _name = "base.image.importer"
+    _inherit = [
+        "base.importer",
+    ]
+    _usage = "image.importer"
+
+    def _download_image(self, url, timeout=None):
+        headers = {
+            "User-Agent": USER_AGENT,
+        }
+
+        try:
+            response = requests.get(url, headers=headers, timeout=timeout)
+        except requests.exceptions.Timeout:
+            raise NetworkRetryableError("Timeout while downloading image")
+        return response.content if response.ok else None
+
+    def run(self, url, binding_model, record_id, image_field, timeout=None):
+        image = self._download_image(url, timeout=timeout)
+        if not image:
+            return "No image could be downloaded"
```

Why it just returns some string, whether it imported or not? Dont we need a way to know if it was successful or not, like True/False?

---

The string returned from a job method [is stored](https://github.com/OCA/queue/blob/af6193f393404e4361f6f43f517e1ff6a688b54e/queue_job/job.py#L517) on the job record and can we seen on the job record in the job form view.

I believe I initially chose to not fail the job if the image can't be downloaded or if it is an invalid image since it is only used for importing user avatars and it did not matter much if it failed and we did not want to have lingering failed jobs for it.

Let me know if we should raise an exception/have failed jobs instead.

---

Not necessarily fail it, but at least we can return some result that informs you whether it failed or not.

Like let say, instead of just string, we would return tuple (bool, str), where `bool` is value indicating whether import was success and another item in tuple, showing some message, like you do now.

When we return just message, its not very good indicator whether import succeeded or not.

If we would let say want to extend this:

with `res = super().run()`, we might want to know result of import easily. And it would be hard to find it out by just message.
So something like

```python
   success, msg = super().run()
   if success:
       # do something
```

---

> When we return just message, its not very good indicator whether import succeeded or not.

@oerp-odoo I agree, however, this was not meant as a failure/success indicator to developers, as mentioned, this is part of `queue_job` API - the job method return value [is stored](https://github.com/OCA/queue/blob/af6193f393404e4361f6f43f517e1ff6a688b54e/queue_job/job.py#L517) in the [`queue.job` record `result` field](https://github.com/OCA/queue/blob/af6193f393404e4361f6f43f517e1ff6a688b54e/queue_job/models/queue_job.py#L99).